YatimaStdLib/AddChain.lean

```diff
@@ -0,0 +1,183 @@
+import YatimaStdLib.Array
+import YatimaStdLib.Nat
+
+/-!
+# AddChain
+
+This module implements an efficient representation of numbers in terms of the double-and-add 
+algorithm.
+
+The finding an `AddChain` with `minChain` can be used to pre-calculate the function which
+implements a double-and-add or square-and-multiply represention of a natural number.
+
+The types are left polymorphic to allow for more efficient implementations of `double` in the 
+case of `Chainable`, or `square` for `Square`.
+
+## References
+
+* Implementation taken from: https://crates.io/crates/addchain
+* Algorithm based on _Addition Chains using Continued Fractions_ by Bergeron, Berstel, Brlek, Duboc
+  (https://www.sciencedirect.com/science/article/abs/pii/0196677489900369)
+-/
+
+/-- The typeclass which implements efficient `add`, `mul`, `double`, and `doubleAdd` methods -/
+class Chainable (α : Type _) extends BEq α, OfNat α (nat_lit 1) where
+  add : α → α → α 
+  mul : α → α → α 
+  double : α → α := fun x => add x x
+  doubleAdd : α → α → α := fun x y => add (double x) y 
+
+/-
+In this section we include the basic instances for `Chainable` that exist in the core 
+numerical libraries of Lean
+-/
+section instances
+
+instance (priority := low) [Add α] [Mul α] [BEq α] [OfNat α (nat_lit 1)] : Chainable α where
+  add := Add.add
+  mul := Mul.mul
+
+instance [Chainable α] : Inhabited α where
+  default := 1
+
+instance [Chainable α] : HAdd α α α where
+  hAdd := Chainable.add
+
+instance [Chainable α] : HMul α α α where
+  hMul := Chainable.mul
+
+end instances
+
+/-- The steps that can appear in an `AddChain` -/
+inductive ChainStep | add (idx₁ idx₂ : Nat) | double (idx : Nat)
+deriving Repr
+
+/--
+The chain of operations that can be used to represent a `Chainable` n in terms of the 
+double-and-add algorithm
+-/
+def AddChain (α : Type _) [Chainable α] := Array α
+
+instance [Chainable α] : Inhabited (AddChain α) where
+  default := #[1]
+
+instance [Chainable α] : HAdd (AddChain α) α (AddChain α) where
+  hAdd ch n := ch.push (n + ch.back)
+
+instance [Chainable α] : Mul (AddChain α) where
+  mul ch₁ ch₂ := 
+    let last := ch₁.back
+    let ch₂' := ch₂.last.map (fun x => x * last) 
+    ch₁.append ch₂'
+
+/- 
+In this section we implement an efficient algorithm to calculate the minimal AddChain for a natural 
+number
+-/
+section nat
```

Comment:
Why start a section here?

Reply:
You're right, the extra section is superfluous with the namespace